04-onion-routing.md

```diff
@@ -754,58 +754,141 @@ Note that these error messages are not reliable, as they are not placed on-chain
 due to the possibility of hop failure.
 
 Intermediate hops store the shared secret from the forward path and reuse it to
-obfuscate any corresponding return packet during each hop.
+authenticate and obfuscate any corresponding return packet during each hop.
 In addition, each node locally stores data regarding its own sending peer in the
 route, so it knows where to return-forward any eventual return packets.
-The node generating the error message (_erring node_) builds a return packet
+
+The origin node signals to each node on the path that it supports the error
+format that is detailed below via a tlv record in the forward packet [TO BE
+SPECIFIED].
+
+## Erring node
+
+The node generating the error message builds a return packet
 consisting of the following fields:
 
 1. data:
-   * [`32*byte`:`hmac`]
    * [`u16`:`failure_len`]
    * [`failure_len*byte`:`failuremsg`]
    * [`u16`:`pad_len`]
    * [`pad_len*byte`:`pad`]
+   * [`27*9*byte`:`payloads`]
+   * [`378*32*byte`:`hmacs`]
+
+Where `failuremsg` is defined below, and `pad` are the extra bytes used to
+conceal length.
+
+The field `payloads` contains a per-hop payload. The erring node puts its
+payload at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum number of hops in a route (27) and the payload
+size (9 bytes).
+
+The per-hop payload consists of the following fields:
+1. data:
+   * [`byte`:`payload_type`]
+   * [`u64`:`hold_time_ms`]
+
+Where `payload_type` indicates whether this hop is an intermediate hop (value 0)
+or the source of the error (value 1). The erring node sets this to 1. Via
+`hold_time_ms` each hops reports the time that it held on to the htlc. The
```

Comment:
If nodes can expect to be penalized based on this, what is keeping them from always reporting a bogus hold time?

Reply:
They can, but if their peers do report the true hold time, they won't escape penalization.